extension/observer/dockerobserver/extension.go

```diff
@@ -84,29 +86,89 @@ func (d *dockerObserver) Shutdown(ctx context.Context) error {
 
 // ListAndWatch provides initial state sync as well as change notification.
 // Emits initial list of endpoints loaded upon extension Start. It then goes into
-// a loop to sync the container cache periodically and change endpoints.
-// TODO: Watch docker events to notify listener of changed endpoints as
-// events stream
+// a loop to sync the container cache periodically, watch for docker events,
+// and notify the observer listener of endpoint changes.
 func (d *dockerObserver) ListAndWatch(listener observer.Notify) {
 	d.emitContainerEndpoints(listener)
-
 	go func() {
-		ticker := time.NewTicker(d.config.CacheSyncInterval)
-		defer ticker.Stop()
+		cacheRefreshTicker := time.NewTicker(d.config.CacheSyncInterval)
+		defer cacheRefreshTicker.Stop()
+		// timestamp to begin looking back for events
+		lastTime := time.Now()
+		var eventErr error
+
+		// loop for re-connecting to events channel if a client error occurs
+		for {
```

Comment:
Does this need to be split into another for loop? The previously linked internal docker ContainerEventLoop has reconnect logic in the error case so that managing loop state elsewhere is unnecessary.

Reply:
This was an attempt to remove the need for the loop label that's in the original (EVENT_LOOP) after Dmitrii suggested a refactor. In general, `dClient.Events()` needs to be re-invoked if we lose a connection, and I think this satisfies that. Are you suggesting modifying this further?

Reply:
The separation of tasks seems good to me now (event gathering and retrying), thanks.